### PR TITLE
Update go-grpc-middleware to v2.0.0

### PR DESCRIPTION
**Pull request by coleenquadros**
- Update go_grpc_middleware to v2.0.0.
- Remove Tags Interceptor from Thanos. Tags interceptor is removed from v2.0.0 go-grpc-middleware. 
- Tracing interceptor directory from go_grpc_middleware(v2.0.0 rc 2,3,4) moved to Thanos(v2.0.0 makes use of otel). Tracing interceptor was removed in version v2.0.0. The follow up in Thanos would be to migrate to otel.
- Logging middleware migrated to use v2.0.0 logging interceptors. Trace ID is now logged as a part of GRPC request logging(output below). 
(PR for adding trace ID to HTTP logging middleware - https://github.com/thanos-io/thanos/pull/6611)

```
ts=2023-10-20T11:34:12.98939Z caller=grpc.go:107 level=debug name=store service=gRPC/server component=store msg="started call" traceID=5ee8eb9e1b05437670f2d311252aa755 requestID=01HD6DTJQXHWXEHBS7YE7QVJ22 protocol=grpc grpc.component=server grpc.service=thanos.info.Info grpc.method=Info grpc.method_type=unary peer.address=127.0.0.1:52802 grpc.start_time=2023-10-20T13:34:12+02:00 grpc.request.deadline=2023-10-20T13:34:17+02:00 grpc.time_ms=0.145
ts=2023-10-20T11:34:12.989388Z caller=grpc.go:107 level=debug name=store service=gRPC/server component=store msg="started call" traceID=d25ad23a02c034c84fc856a4d6a59fdb requestID=01HD6DTJQX368WC3KT5R45H4WY protocol=grpc grpc.component=server grpc.service=thanos.info.Info grpc.method=Info grpc.method_type=unary peer.address=127.0.0.1:52801 grpc.start_time=2023-10-20T13:34:12+02:00 grpc.request.deadline=2023-10-20T13:34:17+02:00 grpc.time_ms=0.137
ts=2023-10-20T11:34:12.989441Z caller=grpc.go:107 level=debug name=store service=gRPC/server component=store msg="finished call" traceID=5ee8eb9e1b05437670f2d311252aa755 requestID=01HD6DTJQXHWXEHBS7YE7QVJ22 protocol=grpc grpc.component=server grpc.service=thanos.info.Info grpc.method=Info grpc.method_type=unary peer.address=127.0.0.1:52802 grpc.start_time=2023-10-20T13:34:12+02:00 grpc.request.deadline=2023-10-20T13:34:17+02:00 grpc.code=OK grpc.time_ms=0.2
ts=2023-10-20T11:34:12.989445Z caller=grpc.go:107 level=debug name=store service=gRPC/server component=store msg="finished call" traceID=d25ad23a02c034c84fc856a4d6a59fdb requestID=01HD6DTJQX368WC3KT5R45H4WY protocol=grpc grpc.component=server grpc.service=thanos.info.Info grpc.method=Info grpc.method_type=unary peer.address=127.0.0.1:52801 grpc.start_time=2023-10-20T13:34:12+02:00 grpc.request.deadline=2023-10-20T13:34:17+02:00 grpc.code=OK grpc.time_ms=0.199

```

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/logging/options.go
Comment (moadz) [inline pkg/logging/options.go:195]: Comments need removing

File: pkg/tracing/interceptors/reporter.go
Comment (moadz) [inline pkg/tracing/interceptors/reporter.go:1]: Just for clarification: 

We are doing this temporarily until we migrate Thanos Tracing to OTEL provider, instead of OpenTracing. I do believe most of them are OTEL compatible now, but it was a good idea to split it up.

**Review by moadz — APPROVED** [review]
LGTM, thanks for this :)

**Comment by moadz** [comment]
@yeya24, @matej-g any chance we could get a review to try to unblock `go-grpc-middleware` upgrade?

File: pkg/tracing/tracing_middleware/server.go
Comment (douglascamata) [inline pkg/tracing/tracing_middleware/server.go:38]: Should anything be done regarding this commented code here?

**Review by douglascamata — COMMENTED** [review]
Some comments from a 1st pass

File: CHANGELOG.md
Comment (yeya24) [reply CHANGELOG.md:29]: Agree, mentioning user facing changes should be good enough.

File: pkg/tracing/interceptors/client.go
Comment (yeya24) [inline pkg/tracing/interceptors/client.go:6]: Just want to double check, we will remove those vendored code after Thanos fully migrated to OTEL based tracing?

Comment (coleenquadros) [reply pkg/tracing/interceptors/client.go:6]: Yes, thats the intention.

**Review by douglascamata — APPROVED** [review]
LGTM ✅ :shipit:

File: CONTRIBUTING.md
Comment (bwplotka) [reply CONTRIBUTING.md]: Still not resolved?

File: CHANGELOG.md
Comment (bwplotka) [reply CHANGELOG.md:26]: Also, why not adding it, should be simple.

File: go.mod
Comment (bwplotka) [reply go.mod:128]: It might be some dependency that uses old grpc-middleware (likely)

**Comment by yeya24** [comment]
It would be great if we can move this forward. I am kind of blocked by this right now.

**Review by yeya24 — APPROVED** [review]
LGTM. I tried this branch locally and fixed conflicts. It worked for me.

**Comment by fpetkovski** [comment]
Looks like unit tests are still failing.

**Comment by coleenquadros** [comment]
The branch is failing due to this test case in the go-grpc-middleware v2.0.0.rc.2 which was copied over - https://github.com/grpc-ecosystem/go-grpc-middleware/blob/880c1749170035e5a54794546dbfe0d45d10a11d/interceptors/tracing/interceptors_test.go#L78

This is failing in the v2.0.0.-rc.2 tag too

`
=== RUN   TestTaggingSuite
    interceptor_suite.go:127: 
        	Error Trace:	/home/samukher/web/thanos/pkg/tracing/tracing_middleware/grpctesting/interceptor_suite.go:127
        	            				/home/samukher/web/thanos/pkg/tracing/tracing_middleware/grpctesting/interceptor_suite.go:93
        	            				/usr/local/go/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	Test:       	TestTaggingSuite
        	Messages:   	must not error on client Dial
panic: test timed out after 15m0s
running tests:
	TestTaggingSuite (15m0s)
`

**Comment by douglascamata** [comment]
@coleenquadros looks like there is a final and very small linting issue left to be fixed -- just the ordering of imports in a file.

**Comment by douglascamata** [comment]
Could we get another round of reviews, please? 🥺

**Review by fpetkovski — DISMISSED** [review]

**Comment by fpetkovski** [comment]
@coleenquadros would you mind rebasing and we can merge this PR.

**Comment by coleenquadros** [comment]
@fpetkovski @pedro-stanaka I have rebased the changes. Sorry for the delay in response.

**Review by pedro-stanaka — APPROVED** [review]
Thx for your work, this will unblock a lot of nice improvements around metric instrumentation for thanos.

**Pull request merged**
